components/raftstore/src/store/peer.rs

```diff
@@ -1327,9 +1327,47 @@ impl Peer {
             ctx.current_time.replace(monotonic_raw_now());
         }
 
-        // The leader can write to disk and replicate to the followers concurrently
-        // For more details, check raft thesis 10.2.1.
         if self.is_leader() {
+            if let Some(hs) = ready.hs() {
+                // Correctness depends on the fact that the leader lease must be suspected before
+                // other followers know the `PrepareMerge` log is committed, i.e. sends msg to others.
+                // Because other followers may complete the merge process, if so, the source region's
+                // leader may get a stale data.
+                //
+                // Check the committed entries from the last committed index + 1 to the latest committed index.
+                let last_committed_index = self.get_store().committed_index();
+                for entry in self
+                    .raft_group
+                    .raft
+                    .raft_log
+                    .slice(last_committed_index + 1, hs.get_commit() + 1, None)
```

Comment:
Why not check `ready.committed_entries` directly?

Reply:
Because I want to reduce the correctness dependency. If checking `ready.committed_entries` directly, it implies the `committed_entries` must have the latest committed entry which is another correctness dependency.

Reply:
Loading the logs again can hurt performance. I think you can trace the index when proposing and prevent read once it's committed.

Reply:
I think it can't hurt performance. In all cases, these new committed logs exist in raft-rs unstable log or in EntryCache because each proposal is added to EntryCache and is removed after it is applied.

Reply:
>... is added to EntryCache and is removed after it is applied.

That's not true. Raftstore will clear cache when timeout.

On the other hand, even they are still in cache, loading logs requires memory allocation and copy.

Reply:
It's true when the peer is a leader.

> On the other hand, even they are still in cache, loading logs requires memory allocation and copy.
Yes, the memory copy is a little waste. I find it's enough to record one index for splitting or merging because both of them change epoch and only the first one can be committed. I will optimize it soon.

Reply:
I change back to `ready.committed_entries`. And I will optimize it later.